File: 6-api-edge-security-waf/1-java/src/main/java/com/starci/wafgateway/EchoController.java
package com.starci.wafgateway;

import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.client.RestClient;

import java.util.Map;

// EchoController — the gateway route. Requests that survive the WAF filter are
// forwarded to the backend echo service, proving clean traffic passes through.
@RestController
public class EchoController {

    private final RestClient restClient;

    public EchoController() {
        String backendUrl = System.getenv().getOrDefault("BACKEND_URL", "http://localhost:4000");
        this.restClient = RestClient.create(backendUrl);
    }

    @GetMapping("/healthz")
    public Map<String, String> health() {
        return Map.of("status", "ok");
    }

    @PostMapping(value = "/api/echo", consumes = MediaType.ALL_VALUE)
    public ResponseEntity<Map<String, Object>> echo(@RequestBody(required = false) Object body) {
        // Forward the vetted payload to the backend and relay its response.
        Object echo = restClient.post()
                .uri("/echo")
                .contentType(MediaType.APPLICATION_JSON)
                .body(body == null ? Map.of() : body)
                .retrieve()
                .body(Object.class);
        return ResponseEntity.ok(Map.of("status", "allowed", "echo", echo));
    }
}

File: 6-api-edge-security-waf/0-typescript/src/echo.controller.ts
import { Controller, Post, Get, Body, HttpCode } from "@nestjs/common"
import axios from "axios"

// EchoController — the gateway route. Requests that survive the WAF middleware
// are forwarded to the BACKEND echo service, proving clean traffic passes
// end-to-end. The gateway itself never trusts the payload; the WAF already
// vetted it before this handler runs.
@Controller()
export class EchoController {
    private readonly backendUrl = process.env.BACKEND_URL ?? "http://localhost:4000"

    @Get("healthz")
    health(): { status: string } {
        return { status: "ok" }
    }

    @Post("api/echo")
    @HttpCode(200) // Clean requests return 200 (uniform contract across all tracks).
    async echo(@Body() body: Record<string, unknown>): Promise<unknown> {
        // Forward the vetted payload to the backend and relay its response.
        const response = await axios.post(`${this.backendUrl}/echo`, body)
        return { status: "allowed", echo: response.data }
    }
}
